bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/transactions/FileRecoveryLog.java

```diff
@@ -0,0 +1,255 @@
+            }
+        } catch (IOException e) {
+            stderr.println("error: failed to acquire lock on recovery log file " + file.toPath());
+        }
+    }
+
+    @Override
+    public void put(TransactionLogRecord trxRecord) {
+        boolean force = !(trxRecord.getTransactionState().equals(RecoveryState.TERMINATED)); // lazy write for done record
+        writeToFile(trxRecord.getTransactionLogRecord(), force);
+        if (checkpointInterval != -1) {
+            ifNeedWriteCheckpoint();
+            numOfPutsSinceLastCheckpoint++;
+        }
+    }
+
+    public Map<String, TransactionLogRecord> getPendingLogs() {
+        Map<String, TransactionLogRecord> pendingTransactions = new HashMap<>();
+        Map<String, TransactionLogRecord> transactionLogs = readLogsFromFile(file);
+        if (transactionLogs == null) {
+            return null;
+        }
+        for (Map.Entry<String, TransactionLogRecord> entry : transactionLogs.entrySet()) {
+            String trxId = entry.getKey();
+            TransactionLogRecord trxRecord = entry.getValue();
+
+            if (!trxRecord.isCompleted()) {
+                pendingTransactions.put(trxId, trxRecord);
+            }
+        }
+        return pendingTransactions;
+    }
+
+    /**
+     * Write a transaction log entry to the recovery log file.
+     *
+     * @param str the log entry to write
+     */
+    private void writeToFile(String str, boolean force) {
+        if (appendChannel == null || !appendChannel.isOpen()) {
+            initAppendChannel(file);
+        }
+        byte[] bytes = str.getBytes();
+        try {
+            appendChannel.write(java.nio.ByteBuffer.wrap(bytes));
+            appendChannel.force(force);
+        } catch (IOException e) {
+            stderr.println("error: failed to write to recovery log file " + file.toPath());
+        }
+    }
+
+
+    private Map<String, TransactionLogRecord> readLogsFromFile(File file) {
+        Map<String, TransactionLogRecord> logMap = new HashMap<>();
+        if (!file.exists() || file.length() == 0) {
+            return null;
+        }
+        if (appendChannel != null) {
+            closeEverything();
+        }
+        try (BufferedReader reader = new BufferedReader(new FileReader(file))) {
+            String line;
+            while ((line = reader.readLine()) != null) {
+                TransactionLogRecord transactionLogRecord = TransactionLogRecord.parseTransactionLogRecord(line);
+                if (transactionLogRecord != null) {
+                    // TODO: add a check here to check whether the record exists in the logMap and new state is a valid
+                    //  state from the current state
+                    logMap.put(transactionLogRecord.transactionId, transactionLogRecord);
+                }
+            }
+        } catch (IOException e) {
+            stderr.println("error: failed to read the recovery log file " + file.toPath());
```

Comment:
Since you are appending "error:" to every error log, perhaps you can use a utility method that takes the message and appends the prefix and prints the error output.

``` java
private static printError(String message) {
    stderr.println(ERROR_PREFIX + message);
}
```

Reply:
fixed in [6d2287d](https://github.com/ballerina-platform/ballerina-lang/pull/42080/commits/6d2287d8bfab130afca749adc90421c95d5fb3bd)